component/loki/process/internal/stages/jsonmerge.go

```diff
@@ -0,0 +1,96 @@
+package stages
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	json "github.com/json-iterator/go"
+	"github.com/prometheus/common/model"
+)
+
+// ErrEmptyValuesJSONMergeStageConfig error returned if the config is empty.
+var ErrEmptyValuesJSONMergeStageConfig = errors.New("values in jsonmerge config cannot be empty")
+var ErrEmptyOutputJSONMergeStageConfig = errors.New("output in jsonmerge config cannot be empty")
+
+// JSONMergeConfig contains the slice of labels to be dropped.
+type JSONMergeConfig struct {
+	Source string   `river:"source,attr"`
+	Values []string `river:"values,attr"`
+	Output string   `river:"output,attr"`
+}
+
+func newJSONMergeStage(logger log.Logger, config JSONMergeConfig) (Stage, error) {
+	if len(config.Values) < 1 {
+		return nil, ErrEmptyValuesJSONMergeStageConfig
+	}
+
+	if config.Output == "" {
+		return nil, ErrEmptyOutputJSONMergeStageConfig
+	}
+
+	return toStage(&jsonmergeStage{
+		config: config,
+		logger: logger,
+	}), nil
+}
+
+type jsonmergeStage struct {
+	config JSONMergeConfig
+	logger log.Logger
+}
+
+// Process implements Stage.
+func (j *jsonmergeStage) Process(labels model.LabelSet, extracted map[string]interface{}, t *time.Time, entry *string) {
+	obj, err := j.extractSourceObject(extracted, entry)
+	if err != nil {
+		if Debug {
+			level.Debug(j.logger).Log("msg", "unable to extract source value", "error", err.Error())
+		}
+	}
+
+	for _, value := range j.config.Values {
+		extractedValue, ok := extracted[value]
+		if !ok {
+			if Debug {
+				level.Debug(j.logger).Log("msg", "unable to find value from extracted map", "value", value)
+			}
+			continue
+		}
+		obj[value] = extractedValue
+	}
+
+	b, err := json.Marshal(obj)
+	if err != nil {
+		if Debug {
+			level.Debug(j.logger).Log("msg", "unable to json marshal merged object", "error", err.Error())
+		}
+		return
+	}
+
+	extracted[j.config.Output] = string(b)
```

Comment:
Hello @tpaschalis, I have some confusion with adding data to the extracted map and using it in the `stage.output`. The output stage requires that the value from the extracted map cannot be a complex structure. Therefore I need to marshal the data before adding it to the extracted map.

However, this results in stringed JSON objects in the output line, for example:
```
{"something":"[1,2,3]","extracted_object":"{\"one\":true}"}
```
Any ideas for working around this issue?

Reply:
So iiuc you're wondering about the stringed `"[1,2,3]"` field that is no longer an array, right?

The reason behind this is that `stage.json` stores [complex objects](https://github.com/grafana/agent/blob/914b965ba30a752682081c12e9a6433e4a4e3bb0/component/loki/process/internal/stages/json.go#L159-L169) back in their marshalled form, so it gets stored as a string from the very start. On the other hand, numbers and booleans get to keep their type.

Using the `json` filter at query time will give you back the non-stringified form, but there's no way to make Loki treat it as a proper array you can index, slice etc. There's been some discussion in allowing such operations at query time but I'm not sure how far it has progressed.

![image](https://github.com/grafana/agent/assets/17771679/592532e2-3027-4cc8-9b29-5f693b6db741)

Unfortunately there's no other workaround for now; but getting stage.json to handle more types natively might be interesting in the future